File: backend/services/stigmer-server/pkg/domain/agent/defaultagent/defaultagent.go
// Package defaultagent resolves the platform default agent — the single
// resolution implementation behind Agent.GetDefault, AgentExecution create,
// and Session create.
//
// The default agent is a platform-level singleton: the agent labeled
// stigmer.ai/default-agent: "true" with visibility_public, seeded in the
// system org and served to callers of every org. Resolution is deliberately
// GLOBAL — GetDefaultAgentRequest.org exists for authorization scoping only
// (see apis/ai/stigmer/agentic/agent/v1/io.proto). It powers the
// session-first UX where a user starts a conversation without picking an
// agent.
//
// This package is the OSS twin of the cloud edition's
// AgentRepo.findDefault() (stigmer-cloud
// backend/services/stigmer-service/.../domain/agentic/agent/repo/), which
// likewise owns the label AND visibility predicate in one named query
// called by the same three handlers. Keep the two in sync.
//
// # Determinism contract (stigmer/stigmer#356)
//
// Multiple labeled agents is a reachable state, not an error: safe label
// rotation applies the new default before retiring the old one (retiring
// first would leave a window with no resolvable default). Among public
// candidates the winner is the one with the LOWEST metadata.id — the
// incumbent. Rationale, in order of what was rejected:
//
//   - First-match (the pre-#356 behavior): the store scan had no ordering,
//     so the served default depended on row insertion order, and the
//     visibility gate could fail on an arbitrary non-public winner while a
//     valid public candidate existed.
//   - Newest-wins: nothing guards the reserved stigmer.ai/* label namespace
//     at write boundaries, so newest-wins would let any newly labeled public
//     agent deterministically capture the platform-wide default.
//     Incumbent-wins keeps the seeded default serving until an operator
//     explicitly removes its label — that removal is the rotation cutover.
//   - Creation time: spec_audit.created_at is not trustworthy as an
//     ordering key. When this resolver was designed,
//     steps.SetAuditFieldsForUpdate reset it on every visibility update,
//     skill push, and schedule trigger (stigmer/stigmer#453, since fixed);
//     rows written before that fix may still carry rewritten timestamps,
//     and audit metadata remains operationally mutable in a way identity
//     is not. metadata.id never changes for the life of the row and is
//     time-ordered for server-generated ULIDs.
package defaultagent

import (
	"context"
	"errors"
	"fmt"

	"github.com/rs/zerolog/log"
	agentv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/agentic/agent/v1"
	"github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/commons/apiresource"
	"github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/commons/apiresource/apiresourcekind"
	"github.com/stigmer/stigmer/backend/libs/go/store"
	"google.golang.org/protobuf/proto"
)

// Label marks the platform default agent; the value must be LabelValue.
const Label = "stigmer.ai/default-agent"

// LabelValue is the only value of Label that marks an agent as the default.
const LabelValue = "true"

// ErrNotConfigured reports that no agent carries the default-agent label.
// Call sites map this to codes.NotFound with their own caller-facing copy.
var ErrNotConfigured = errors.New("no agent labeled " + Label + "=" + LabelValue)

// ErrNotPublic reports that labeled agents exist but none is
// visibility_public. Call sites map this to codes.FailedPrecondition —
// a deliberate divergence from the cloud edition, whose SQL predicate
// collapses this state into "not found": the distinct code tells a
// self-hosting operator the label is present but the visibility is wrong.
var ErrNotPublic = errors.New("agents labeled " + Label + "=" + LabelValue + " exist but none is visibility_public")

// Find resolves the platform default agent per the package contract:
// candidates are all agents labeled Label=LabelValue, only visibility_public
// candidates are eligible, and among those the lowest metadata.id wins.
//
// Returns ErrNotConfigured when nothing carries the label, ErrNotPublic when
// labeled agents exist but none is public, and a wrapped store error on
// query or decode failure (a store failure is an internal fault, not a
// "no default agent" condition — callers must not map it to NotFound).
func Find(ctx context.Context, s store.Store) (*agentv1.Agent, error) {
	raws, err := s.FindAllByLabel(ctx, apiresourcekind.ApiResourceKind_agent, Label, LabelValue, &agentv1.Agent{})
	if err != nil {
		return nil, fmt.Errorf("list agents labeled %s=%s: %w", Label, LabelValue, err)
	}
	if len(raws) == 0 {
		return nil, ErrNotConfigured
	}

	publicCount := 0
	var winner *agentv1.Agent
	for _, raw := range raws {
		candidate := &agentv1.Agent{}
		if err := proto.Unmarshal(raw, candidate); err != nil {
			// FindAllByLabel already unmarshaled every returned row to match
			// the label, so a failure here is store corruption. Fail loudly:
			// skipping a row could silently change which agent wins.
			return nil, fmt.Errorf("unmarshal agent labeled %s=%s: %w", Label, LabelValue, err)
		}
		if candidate.GetMetadata().GetVisibility() != apiresource.ApiResourceVisibility_visibility_public {
			continue
		}
		publicCount++
		if winner == nil || candidate.GetMetadata().GetId() < winner.GetMetadata().GetId() {
			winner = candidate
		}
	}
	if winner == nil {
		return nil, ErrNotPublic
	}

	if publicCount > 1 {
		// Expected briefly mid-rotation; persistent duplicates mean someone
		// forgot to retire the old label after applying the new default.
		log.Warn().
			Int("public_labeled_agents", publicCount).
			Str("winner_id", winner.GetMetadata().GetId()).
			Str("winner_name", winner.GetMetadata().GetName()).
			Msg("Multiple public agents carry the default-agent label; serving the incumbent (lowest id). Retire the stale label to complete rotation.")
	}

	return winner, nil
}

File: backend/services/stigmer-server/src/domain/agent/defaultagent.ts
/**
 * Platform default-agent resolution — ports
 * pkg/domain/agent/defaultagent/defaultagent.go: the single resolution
 * implementation behind Agent.GetDefault, AgentExecution create (#17), and
 * Session create.
 *
 * The default agent is a platform-level singleton: the agent labeled
 * stigmer.ai/default-agent: "true" with visibility_public, seeded in the
 * system org and served to callers of every org. Resolution is deliberately
 * GLOBAL — GetDefaultAgentRequest.org exists for authorization scoping only
 * (see apis/ai/stigmer/agentic/agent/v1/io.proto). It powers the
 * session-first UX where a user starts a conversation without picking an
 * agent. The cloud twin is AgentRepo.findDefault(); keep the two in sync.
 *
 * Determinism contract (stigmer/stigmer#356): multiple labeled agents is a
 * reachable state, not an error — safe label rotation applies the new
 * default before retiring the old one. Among public candidates the winner
 * is the one with the LOWEST metadata.id — the incumbent. Rejected
 * alternatives (recorded in the Go package doc): first-match depends on row
 * insertion order; newest-wins lets any newly labeled public agent capture
 * the platform-wide default (nothing guards the reserved stigmer.ai/*
 * namespace at OSS write boundaries); creation time is not trustworthy as
 * an ordering key (pre-#453 rows may carry rewritten timestamps).
 * metadata.id never changes and is time-ordered for server-generated ULIDs.
 */
import { fromBinary } from "@bufbuild/protobuf";

import { AgentSchema } from "@stigmer/protos/ai/stigmer/agentic/agent/v1/api_pb";
import type { Agent } from "@stigmer/protos/ai/stigmer/agentic/agent/v1/api_pb";
import { ApiResourceVisibility } from "@stigmer/protos/ai/stigmer/commons/apiresource/enum_pb";
import { ApiResourceKind } from "@stigmer/protos/ai/stigmer/commons/apiresource/apiresourcekind/api_resource_kind_pb";

import type { Logger } from "../../boot/logger.js";
import type { Store } from "../../store/interface.js";

/** Marks the platform default agent; the value must be DEFAULT_AGENT_LABEL_VALUE. */
export const DEFAULT_AGENT_LABEL = "stigmer.ai/default-agent";

/** The only value of DEFAULT_AGENT_LABEL that marks an agent as the default. */
export const DEFAULT_AGENT_LABEL_VALUE = "true";

/**
 * No agent carries the default-agent label. Call sites map this to
 * NotFound with their own caller-facing copy (Go ErrNotConfigured).
 */
export class DefaultAgentNotConfiguredError extends Error {
  constructor() {
    super(
      `no agent labeled ${DEFAULT_AGENT_LABEL}=${DEFAULT_AGENT_LABEL_VALUE}`,
    );
    this.name = "DefaultAgentNotConfiguredError";
  }
}

/**
 * Labeled agents exist but none is visibility_public. Call sites map this
 * to FailedPrecondition — a deliberate divergence from the cloud edition,
 * whose SQL predicate collapses this state into "not found": the distinct
 * code tells a self-hosting operator the label is present but the
 * visibility is wrong (Go ErrNotPublic).
 */
export class DefaultAgentNotPublicError extends Error {
  constructor() {
    super(
      `agents labeled ${DEFAULT_AGENT_LABEL}=${DEFAULT_AGENT_LABEL_VALUE} exist but none is visibility_public`,
    );
    this.name = "DefaultAgentNotPublicError";
  }
}

/**
 * Resolves the platform default agent per the module contract: candidates
 * are all agents labeled DEFAULT_AGENT_LABEL=true, only visibility_public
 * candidates are eligible, and among those the lowest metadata.id wins.
 *
 * Throws DefaultAgentNotConfiguredError when nothing carries the label,
 * DefaultAgentNotPublicError when labeled agents exist but none is public,
 * and lets store/decode failures propagate (a store failure is an internal
 * fault, not a "no default agent" condition — callers must not map it to
 * NotFound).
 */
export async function findDefaultAgent(
  store: Store,
  logger: Logger,
): Promise<Agent> {
  const raws = await store.findAllByLabel(
    ApiResourceKind.agent,
    DEFAULT_AGENT_LABEL,
    DEFAULT_AGENT_LABEL_VALUE,
    AgentSchema,
  );
  if (raws.length === 0) {
    throw new DefaultAgentNotConfiguredError();
  }

  let publicCount = 0;
  let winner: Agent | undefined;
  for (const raw of raws) {
    // findAllByLabel already unmarshaled every returned row to match the
    // label, so a decode failure here is store corruption. Fail loudly
    // (propagate): skipping a row could silently change which agent wins.
    const candidate = fromBinary(AgentSchema, raw);
    if (
      candidate.metadata?.visibility !==
      ApiResourceVisibility.visibility_public
    ) {
      continue;
    }
    publicCount++;
    if (
      winner === undefined ||
      (candidate.metadata?.id ?? "") < (winner.metadata?.id ?? "")
    ) {
      winner = candidate;
    }
  }
  if (winner === undefined) {
    throw new DefaultAgentNotPublicError();
  }

  if (publicCount > 1) {
    // Expected briefly mid-rotation; persistent duplicates mean someone
    // forgot to retire the old label after applying the new default.
    logger.warn(
      "Multiple public agents carry the default-agent label; serving the incumbent (lowest id). Retire the stale label to complete rotation.",
      {
        publicLabeledAgents: publicCount,
        winnerId: winner.metadata?.id ?? "",
        winnerName: winner.metadata?.name ?? "",
      },
    );
  }

  return winner;
}
